Implements a stack of dispatched events #7404

burzum opened the pull request:
See #7404

Commit:
Implements a stack of dispatched events #7404

markstory commented on src/Event/EventManager.php:
```diff
@@ -361,6 +369,7 @@ public function dispatch($event)
                 $event->result = $result;
             }
         }
+        $this->_dispatchedEvents[] = $event;
```
This could end up consuming a non-trivial amount of memory. This needs to be an opt-in feature.

burzum replied:
How would you like to en/disable it? Making it a dev only feature by checking for debug = 1? Having a method like enableEventStacking() and disableEventStacking()? I'm suggesting two methods because I think in one of the other issue tickets I've seen that we want to feature getters and setters (ugly :() in the future.

markstory replied:
Having `disable/enableEventTracking()` sounds reasonable to me.

Commit:
Disable event stacking by default.

This needs to be explicitly enable because otherwise it could cause memory issues for people without that they'll know the cause.

burzum commented:
@markstory if you see my last commit, I think we could move all of these new methods into a new class for the sake of SoC.

This would also allow people, like the Croogo guys, to implement their own event stack for whatever reason. Any wish / idea how the interface for this should look like?

EventManager::setEventStack(EventStack $stack) maybe?

markstory commented on tests/TestCase/Event/EventManagerTest.php:
```diff
@@ -710,4 +710,40 @@ public function testDispatchWithGlobalAndLocalEvents()
+     *
+     * @return void
+     * @triggers my_event $this)
+     * @triggers my_second_event $this)
```
What is with the `)` here?

markstory commented on src/Event/EventManager.php:
```diff
@@ -461,6 +478,60 @@ public function matchingListeners($eventKeyPattern)
+     *
+     * @return void
+     */
+    public function disableEventStacking()
```
Isn't this 'tracking' dispatched events?

burzum replied:
Guess it's a matter of interpretation. You build up a stack (to stack) them or you can for sure say as well "I'm tracking" them. I'll change it.

ADmad replied:
I too think `disable/enableEventTracking()` are more suitable method names.

burzum replied:
OK, I'll rename them tomorrow, I'm about to leave for BBQ. :meat_on_bone:

markstory commented:
> This would also allow people, like the Croogo guys, to implement their own event stack for whatever reason. Any wish / idea how the interface for this should look like?

I had a similar thought and then deleted it, as I thought it would be over-engineered. My thought was:

```php
$eventList = [];
EventManager::trackEvents(function ($event) use (&$eventList) {
  $eventList[] = $event;
});
```

This would let people who are interested in tracking events do so through a callable.

Commit:
phpcs fix

burzum commented:
@markstory Why do you think this is over-engineered?

I think there the list of requirements we have is this:

* Event tracking / stacking should be optional, by default it's off due to memory usage
* You should be able to flush the stack at any time
* The user should be able to do something with the stack at *any* time

So what about implementing an EventStack and allow devs to set the class or instance of stack object they would like to use? Like `$eventManager->eventStack()->someMethod()`

This would be proper Soc, give the devs the ability to alter the behavior of the stack and still be an optional thing. To keep the API flat we could change enableEventStacking() take true or an EventStack object or a string to a class that will be instantiated and used as an event stack.

markstory commented:
The separate object approach looks nice. If you wanted a low-effort implementation [SplStack](http://php.net/splstack) might be a decent way to track/stack events that were triggered.

Commit:
Moving the event stack into a separate object.

markstory commented on src/Event/EventStack.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    public function add(Event $event)
+    {
+        $this->_events[] = $event;
```
This isn't _really_ a 'stack'. A stack usually prepends into the underlying list.

markstory commented on src/Event/EventStack.php:
```diff
@@ -0,0 +1,136 @@
+     * An offset to check for.
+     * </p>
+     * @return boolean true on success or false on failure.
+     * </p>
```
You don't need HTML tags in the doc blocks. The API generation tool will take care of this.

burzum replied:
My IDE added them automatically, they're now fixed.

markstory commented on src/Event/EventStack.php:
```diff
@@ -0,0 +1,136 @@
+     * </p>
+     * <p>
+     * The return value will be casted to boolean if non-boolean was returned.
+     * @since 5.0.0
```
Not sure about this version number.

burzum replied:
My IDE added them automatically, they're now fixed.

Commit:
Updating doc blocks

markstory commented on src/Event/EventStack.php:
```diff
@@ -0,0 +1,113 @@
+     */
+    public function offsetExists($offset)
+    {
+        return isset($this->_events[$offset]);
```
It would be good to get a test that covers the `ArrayAccess` interface.

Commit:
Adding tests for ArrayAccess for the EvenStack

lorenzo commented:
@burzum What do you plan to use this for?

I'm concerned that always stacking the events will prevent PHP from reclaiming the memory from the objects that are stored in the event, potentially leading to memory leaks.

bcrowe commented on src/Event/EventManager.php:
```diff
@@ -461,6 +471,50 @@ public function matchingListeners($eventKeyPattern)
+     * @param \Cake\Event\Event $event An event to add to the stack.
+     * @return void
+     */
+    public function _stackEvent(Event $event)
```
`public` but has `_` prefix?

burzum replied:
Fixed.

burzum commented:
@lorenzo this was discussed before, the stacking is optional, you'll have to explicitly enable it for exactly this reason.

jippi commented:
CRUD does a similar thing - https://github.com/FriendsOfCake/crud/blob/master/src/Event/Subject.php#L20

Commit:
API improvements for the event stack and tests.

burzum commented:
@jippi I've added the hasEvent() method as well, I think this one is pretty useful.

@markstory we will use getters and setters in the future, right? So how about this?

* EventManager::setEventStack() - adds the stack object
* EventManager::getEventStack() - gets the stack object
* EventManager::unsetEventStack() - removes the stack object and disables event stacking

ADmad commented:
As @markstory pointed in one of the comments above this isn't really a stack. It's a list of events fired. So I would suggest replacing "stack" with "list" in all method / class names.

Commit:
Renaming the event stack files to event list.

Commit:
Renaming event stack to event list.

Commit:
Renaming some methods to access the event list.

burzum commented:
@markstory ping? Anything else left?

ADmad commented on src/Event/EventManager.php:
```diff
@@ -54,6 +54,20 @@ class EventManager
+    protected $_eventList;
+
+    /**
+     * Enables automatic adding of events to the event stack object if it is present.
```
s/stack/list

ADmad commented on src/Event/EventManager.php:
```diff
@@ -54,6 +54,20 @@ class EventManager
+     *
+     * @param bool
+     */
+    protected $_listEvents = false;
```
Wouldn't `$_trackEvents` be a better name?

ADmad commented on src/Event/EventManager.php:
```diff
@@ -461,6 +482,63 @@ public function matchingListeners($eventKeyPattern)
+     * @param \Cake\Event\Event $event An event to add to the stack.
+     * @return void
+     */
+    public function listEvent(Event $event)
```
`listEvent` is a funny name.

burzum replied:
Fixed.

ADmad commented on src/Event/EventList.php:
```diff
@@ -0,0 +1,129 @@
+/**
+ * The Event List
+ */
+class EventList implements \ArrayAccess, \Countable
```
Wouldn't `EventsList` (plural) be better?

dereuromark replied:
Usually in English the first word is singular.

ADmad replied:
Huh?

burzum replied:
I don't care guys, just settle one something and tell me.

dereuromark replied:
It is always EventListener, EventDispatcher, EventPlanner or in this case EventList :D

ADmad replied:
@dereuromark That makes lot more sense then your previous comment :)  So current name is fine to maintain naming consistency.

ADmad commented on src/Event/EventManager.php:
```diff
@@ -461,6 +482,63 @@ public function matchingListeners($eventKeyPattern)
     /**
+     * Returns the event list.
+     *
+     * @return array
```
Return type is incorrect.

burzum replied:
Fixed

Commit:
Renaming a few methods, doc block improvements

Commit:
Fixing the return type in a doc block

Commit:
Fixing typo / copy paste error in doc blocks
